include/nanobind/nb_class.h

```diff
@@ -354,6 +364,13 @@ class class_ : public object {
         if constexpr (!std::is_same_v<Base, T>) {
             d.base = &typeid(Base);
             d.flags |= (uint32_t) detail::type_init_flags::has_base;
+
+            if (uintptr_t offset = (uintptr_t) (Base*) (T*) 0x1000 - 0x1000) {
```

Comment:
All of the other operations in `class_::class_` are  `noexcept`, which an intentional choice to avoid C++ compilers generating unwind tables for most binding code. So putting `detail::raise` in here seems a little concerning (even if "disabled" by the condition in virtually all cases). Could there be a way of capturing this in a `constexpr` fashion?

Reply:
Unfortunately I'm not aware of any reasonable way to check the offset of a base class subobject at compile time. There is an unreasonable way:
```
template <class Container> struct offset_helper {
  private:
    union U {
        U() {}
        ~U() {}
        char c[sizeof(Container)];
        Container o;
    };
    static U u;
#ifdef __clang__
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wundefined-var-template"
#endif
    static constexpr size_t offset(const void* loc) {
#if defined(__clang__) || __GNUC_GE__(11, 0)
        for (size_t i = 0; i != sizeof(Container); ++i) {
            if (((void const*) &(u.c[i])) == loc) {
                return i;
            }
        }
        throw "unable to compute offset of member";
#else
        return (size_t) loc - (size_t) & (u.o);
#endif
    }
  public:
    template <class Base> static constexpr size_t base_offset() {
        return offset(static_cast<const Base*>(&u.o));
    }
#ifdef __clang__
#pragma clang diagnostic pop
#endif
};
```
But I think this is too heavyweight for nanobind; for example, it takes compilation time that's linear in the size of the bound class (although I haven't measured the constant factor).

`offsetof` produces a constant expression, and as of C++17 it conditionally supports non-standard-layout classes, but I don't know of any way to get it to locate a base class subobject if you don't know a name within that class.

Would changing the `raise` to a `fail` meet your needs here?